pkgs/stdenv/generic/make-derivation.nix

```
@@ -709,7 +709,9 @@ let
       );
 
     let
-      envIsExportable = isAttrs env && !isDerivation env;
+      mainProgram = env.mainProgram or meta.mainProgram or attrs.pname or null;
```

```suggestion
      mainProgram = meta.mainProgram or attrs.pname or null;
```

No derivation today sets `env.mainProgram`, and we don't want to have that as a different or higher-priority value than `meta.mainProgram`, I would think.

---

Also we should not default it to `pname`. That logic can stay in the hook (and emit a warning so we can eventually deprecate it).

---

@emilazy I fully agree, in fact already I've already removed [or `attrs.pname`](https://github.com/bengsparks/nixpkgs/blob/24dc1ea8c3732ce14317935c823bbc5ac3d5e296/pkgs/stdenv/generic/make-derivation.nix#L719) and kept the reference to `pname` [in the hook](https://github.com/bengsparks/nixpkgs/blob/b5a503795141ad7f47c0a2a9d21d90841ea8762a/pkgs/by-name/ve/versionCheckHook/hook.sh#L45) 😄 As mentioned above, I've avoided pushing so far due to eval simply failing, but I look forward to pushing these updates soon.

`NIX_MAIN_PROGRAM` does indeed sound and read nicer. I'll add the warning with a similar message as in `lib.getExe`

---

In fact, I'll push these changes now, so reviews can take them into account.